src/AppInstallerCLICore/Workflows/PinFlow.cpp

```diff
@@ -0,0 +1,231 @@
+
+        if (existingPin)
+        {
+            // Pin already exists.
+            // If it is the same, we do nothing. If it is different, check for the --force arg
+            if (pin == existingPin)
+            {
+                AICLI_LOG(CLI, Info, << "Pin already exists");
+                context.Reporter.Info() << Resource::String::PinAlreadyExists << std::endl;
+                return;
+            }
+
+            AICLI_LOG(CLI, Info, << "Another pin already exists for the package");
+            if (context.Args.Contains(Execution::Args::Type::Force))
+            {
+                AICLI_LOG(CLI, Info, << "Overwriting pin due to --force argument");
+                context.Reporter.Warn() << Resource::String::PinExistsOverwriting << std::endl;
+                pinningIndex->UpdatePin(pin);
+            }
+            else
+            {
+                context.Reporter.Error() << Resource::String::PinExistsUseForceArg << std::endl;
+                AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_PIN_ALREADY_EXISTS);
+            }
+        }
+        else
+        {
+            pinningIndex->AddPin(pin);
+            AICLI_LOG(CLI, Info, << "Finished adding pin");
+            context.Reporter.Info() << Resource::String::PinAdded << std::endl;
+        }
+    }
+
+    void RemovePin(Execution::Context& context)
+    {
+        auto package = context.Get<Execution::Data::Package>();
+        std::vector<Pinning::Pin> pins;
+
+        // TODO: We should support querying the multiple sources for a package, instead of just one
+        auto availableVersion = package->GetLatestAvailableVersion();
+
+        auto pinningIndex = context.Get<Execution::Data::PinningIndex>();
+        Pinning::PinKey pinKey{
+            availableVersion->GetProperty(PackageVersionProperty::Id).get(),
+            availableVersion->GetProperty(PackageVersionProperty::SourceIdentifier).get() };
+        AICLI_LOG(CLI, Info, << "Removing pin for package [" << pinKey.PackageId << "] from source [" << pinKey.SourceId << "]");
+        if (!pinningIndex->GetPin(pinKey))
+        {
+            AICLI_LOG(CLI, Warning, << "Pin does not exist");
+            context.Reporter.Warn() << Resource::String::PinDoesNotExist(pinKey.PackageId) << std::endl;
+            AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_PIN_DOES_NOT_EXIST);
+        }
+
+        pinningIndex->RemovePin(pinKey);
+    }
+
+    void ReportPins(Execution::Context& context)
+    {
+        const auto& pins = context.Get<Execution::Data::Pins>();
+        if (pins.empty())
+        {
+            context.Reporter.Info() << Resource::String::PinNoPinsExist << std::endl;
+            return;
+        }
+
+        // TODO: Use package and source names
+        Execution::TableOutput<4> table(context.Reporter,
+            {
+                Resource::String::SearchId,
+                Resource::String::SearchSource,
+                Resource::String::SearchVersion,

```

Comment:
>Resource::String::SearchVersion, [](http://example.com/codeflow?start=16&length=32)

Do you think the column name "Version" should be more specific to indicate it's a gated version?

Reply:
I don't think we need it. I believe a user would easily understand what we mean if the header is "Version" and the value is something like "1.2.*". I also didn't actually use "gated version" in any of the resource strings, just in logs and names in the code, so it's not something they'd have seen elsewhere.